Refactoring '''get_extractor''' in capa/main.py

aaronatp opened the pull request:
<!--
Thank you for contributing to capa! <3

Please read capa's CONTRIBUTING guide if you haven't done so already.
It contains helpful information about how to contribute to capa. Check https://github.com/mandiant/capa/blob/master/.github/CONTRIBUTING.md

Please describe the changes in this pull request (PR). Include your motivation and context to help us review.

Please mention the issue your PR addresses (if any):
closes #issue_number
-->


### Checklist

<!-- CHANGELOG.md has a `master (unreleased)` section. Please add bug fixes, new features, breaking changes and anything else you think is worthwhile mentioning in the release notes to this file. -->
- [x] No CHANGELOG update needed
<!-- Tests prove that your fix/work as expected and ensure it doesn't break on the feature. -->
- [x] No new tests needed
<!-- Please help us keeping capa documentation up-to-date -->
- [x] No documentation update needed

These edits address issue #1813. 

The edits are meant to ensure that '''get_extractor''' does not perform more than one significant function. In addition, I hope these edits make the logic of the '''get_extractor''' function more clear. With the edits, it is apparent that the function: 1) performs some preliminary error checking; 2) checks whether the '''format_''' or '''backend''' match any existing extractor handling routines; and, 3) if not, raises an error.

Unfortunately, I pushed these changes too early and am now unable to make further edits. In lines 597, "format" should be changed to "format_." In addition, there are three blank lines between some functions - this is inconsistent with the other code in the capa/main.py file, and because of PEP8, these three blank lines should be replaced with two blank lines. I believe these edits are otherwise compliant with the CONTRIBUTING.md guide. Please let me know if I have overlooked something.

I do not believe I have a PR address.

If there are any other questions or concerns about these changes, I would be happy to enhance these edits further.

aaronatp reviewed (commented):
This code fails the "code_style" test for these reasons:

"capa/main.py:546:16: F401 `binaryninja` imported but unused; consider using `importlib.util.find_spec` to test for availability
capa/main.py:547:33: F401 `binaryninja.BinaryView` imported but unused; consider using `importlib.util.find_spec` to test for availability
capa/main.py:561:13: F821 Undefined name `BinaryView`
capa/main.py:561:26: F821 Undefined name `binaryninja`
Found 4 errors."

These errors seem to refer to the "import_binja" function, which imports "binaryninja" and "BinaryView"; and the "handle_binja_backend" function, which uses "binaryninja" and "BinaryView." The test seems to be intended to prevent the use of a function that is not for-sure imported. But the function that _uses_ "binaryninja" and "BinaryView"  is only called by the function that _imports_ "binaryninja" and "BinaryView"; and the function that _imports_ "binaryview" and "BinaryView" is _always_ called by the function that uses those modules. Therefore, even though this code fails the "code_style" test, it doesn't seem like this is necessarily an issue.

mr-tz commented:
Some changes here look good. The issue was initially meant to simplify the invocation and handling of samples around
- identifying the format and
- getting an extractor

Do you see potential for those and would you also consider https://github.com/mandiant/capa/issues/1821 while taking a look into this?

mr-tz commented on capa/main.py:
```diff
@@ -533,74 +611,23 @@ def get_extractor(
-
-        if os_ == OS_AUTO and not is_supported_os(path):
-            raise UnsupportedOSError()
+        check_supported_format(path, os_)
```
reconsidering, I don't feel like these additional functions necessarily make the code easier to read

especially here where it's not obvious that this may raise an exception...

aaronatp replied:
That is not something I thought about before, but this is helpful to consider. Thank you!

mr-tz commented on capa/main.py:
```diff
@@ -533,74 +611,23 @@ def get_extractor(
-        import capa.features.extractors.dnfile.extractor
-
-        return capa.features.extractors.dnfile.extractor.DnfileFeatureExtractor(path)
+        return handle_dotnet_format(format_)
```
... or here where only a few lines are wrapped

So, I'd vote to leave this function as is.

aaronatp commented:
Hi @mr-tz ,

Thanks for these comments, and for explaining what issue #1813 was meant to achieve.

I think that refactoring some of the '''get_extractor''' format-identifying components could reduce the function's complexity quite a bit. Looking at the original code that specifically handles the binja and viv extractors, the code looks reasonably simple, although I could be wrong. Are there particular parts of the code for identifying the format and getting the extractor that you think could be simplified? Are you referring to the invocation of '''get_extractor''' in the '''main''' function? 

I agree with what @williballenthin indicated in his proposals in #1821, which is that '''get_extractor''' and many of the helper functions that it relies on should probably be moved to a different file. Since '''get_extractor''' and the helper functions it relies on have a relatively specific goal (to return FeatureExtractor objects), it may even be worth moving '''get_extractor''' and its helper functions to a feature_extractor.py file. Are there other things about #1821 you think may be worth considering?

In addition, it may be a good idea to move the try-except statement in viv-extractor code to a new function. While the try-except clause is not especially complex, I found it a bit visually unwieldy, especially considering that the clause is within both a "with" and "if-else" statement.

Also, thank you for pointing out that the names of error handling functions should make clear that they may raise exceptions. This is not something that I had considered before, but I certainly see the value. I have changed the name of the '''is_supported_format''' function to '''check_unsupported_raise_exception''' to reflect this. I have also changed the name of '''import_binja''' to '''attempt_binja_import''', and created '''attempt_save_workspace''', to reflect the error handling abilities of these functions.

Please let me know if you have any more questions or comments about these edits, or if you think I should make further enhancements.

Best,

Aaron

mr-tz commented on capa/main.py:
```diff
@@ -533,74 +605,25 @@ def get_extractor(
-
-        if os_ == OS_AUTO and not is_supported_os(path):
-            raise UnsupportedOSError()
+        check_unsupported_raise_exception(path, os_)
```
I'd propose to just leave the code verbatim here instead of in a new function. Or do you see much benefit added by the function?

mr-tz commented on capa/main.py:
```diff
@@ -517,6 +517,78 @@ def get_workspace(path: Path, format_: str, sigpaths: List[Path]):
+def handle_binja_backend(path: Path, disable_progress: bool) -> FeatureExtractor:
+    import capa.features.extractors.binja.extractor
+
+    attempt_binja_import()
```
I think this can also stay in here and likely resolves the ruff error?

mr-tz commented on capa/main.py:
```diff
@@ -517,6 +517,78 @@ def get_workspace(path: Path, format_: str, sigpaths: List[Path]):
+
+        if should_save_workspace:
+            logger.debug("saving workspace")
+            attempt_save_workspace(vw)
```
this is also simple enough to just leave here

mr-tz commented:
> Are you referring to the invocation of '''get_extractor''' in the '''main''' function?

Yes, that's what I meant originally. Can we wrap the calling and error handling to get an extractor?

> Are there other things about https://github.com/mandiant/capa/issues/1821 you think may be worth considering?

I'd propose to start with what's listed in `proposal 3` and we can further split the functions up if needed.


Overall, I still think we don't need to split the function in so many sub parts but would be very happy to see the refactor into a separate module here.

mr-tz commented:
Nice idea, I think a decorator could work well here. I also like the potential simplification around the `log_unsupported_*` functions.

Let's focus this PR on these changes (rollback the first 5 commits - for now) and then work on changes related to #1821 separately as they're not directly related. Sorry for the confusion :)

mr-tz commented on capa/helpers.py:
can these go into the same list of tuples?

mr-tz commented on capa/helpers.py:
let's use snake case names and maybe rename this to `catch_log_return_errors` or similar?

then let's use via `@<decorator_name>` (see, e.g., https://rinaarts.com/declutter-python-code-with-error-handling-decorators/)

aaronatp commented:
Hi @mr-tz thanks for this feedback! I have made the changes you suggested. Please let me know if you think the list unpacking looks good. I am unsure if the list unpacking is very readable with the message tuples included. I have put the decorator above the function definition. I don't know if it's possible to place the decorator above the function invocation and also assign the function output to a variable - it seems like this would be good because the potential error handling would be visible in the '''main''' function.

Also, no worries regarding #1821! I will rollback the first 5 commits. Thank you for your feedback so far! Please let me know if you have any more comments or suggestions!

aaronatp commented:
Hi @mr-tz I reverted the first 5 or so commits - is this how I should roll back those commits? I am not sure if my reverting worked. If not, I was considering deleting this branch and creating a fresh one with just the most recent commits. Do you think that would be a sensible option?

gmr-tz commented:
Yes, I think creating a new branch may be the easier option at this point.

aaronatp commented:
@mr-tz great will do, thank you.

mr-tz commented:
Thanks, #1845 supersedes this so feel free to close it here.
On first glance the new PR looks good, will review in more detail later.

aaronatp commented:
@mr-tz sounds good, will do! Thank you!
